src/main/java/net/jonathangiles/tools/teenyhttpd/json/TeenyJson.java

```diff
@@ -0,0 +1,352 @@
+        baos.write(writeValueAsString(value).getBytes());
+    }
+
+    /**
+     * Read a JSON string and return a collection of objects of the given type.
+     *
+     * @param collectionType one of List or Set
+     * @param json the JSON string to parse
+     * @return the parsed object which could be null.
+     */
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    public <T, K> K readCollection(String json, Class<? extends Collection> collectionType, Class<T> type) {
+
+        if (json == null) return null;
+
+        Object result = new JsonDecoder(json)
+                .read();
+
+        if (result == null) {
+            return null;
+        }
+
+        List<?> list = (List<?>) result;
+
+        if (List.class.isAssignableFrom(collectionType)) {
+            return (K) list.stream()
+                    .map(o -> parseObject(o, type))
+                    .collect(Collectors.toList());
+        }
+
+        if (Set.class.isAssignableFrom(collectionType)) {
+            return (K) list.stream()
+                    .map(o -> parseObject(o, type))
+                    .collect(Collectors.toSet());
+        }
+
+        throw new IllegalStateException("Unsupported collection type: " + collectionType.getName());
+    }
+
+    /**
+     *
+     * Read a JSON string and return an object of the given type.
+     *
+     * @param json the JSON string to parse
+     * @param type the type of the object to parse
+     *
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T readValue(String json, Class<T> type) {
+
+        if (json == null) return null;
+
+        if (type == String.class) {
+            return (T) json;
+        }
+
+        Object result = new JsonDecoder(json).read();
+
+        if (result == null) {
+            return null;
+        }
+
+        return parseObject(result, type);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T> T readValue(String json, Type type) {
+
+        if (json == null) return null;
+
+        if (type == String.class) {
+            return (T) json;
+        }
+
+        Object result = new JsonDecoder(json).read();
+
+        if (result == null) {
+            return null;
+        }
+
+        if (type instanceof Class<?>) {
+            return (T) parse(result, type);
+        }
+
+        throw new IllegalStateException("Unsupported type: " + type.getTypeName() + " " + type.getClass().getName());
+    }
+
+
+    /**
+     * Register a custom serializer for the given class.
+     *
+     * @param clazz the class to register
+     */
+    public TeenyJson setSerializer(Class<?> clazz, ValueSerializer serializer) {
+        encoder.setSerializer(clazz, serializer);
+        return this;
+    }
+
+    /**
+     * Register a custom parser for the given class.
+     *
+     * @param clazz the class to register
+     */
+    public TeenyJson setParser(Class<?> clazz, ValueParser<?> parser) {
+        parsers.put(clazz, parser);
+        return this;
+    }
```

Comment:
I wonder if these methods should be named `registerParser` and `registerSerializer`?